### DISCUSSION TITLE
Progress Update on Freenet Development - 3rd May, 2023

### DISCUSSION THREAD

**Discussion by sanity** [freenet/freenet-core#617] [General]
# Progress Update on Freenet Development
## 3rd May, 2023

The Freenet development team has been working on a new version of Freenet, codenamed "Locutus," which aims to serve as a decentralized replacement for the World Wide Web. The team has made significant progress in developing the client API and interface for Locutus, and they have discussed some of the technical details and challenges they have encountered.

Key Highlights:

1. Client API and Interface:
   - The team has developed an interface for Locutus, which is part of the standard library.
   - The client API consists of two main modules: "browser" and "regular." The "browser" module is used for developing WebAssembly applications, such as the email app they are working on. The "regular" module is used for standalone applications that communicate with the node via WebSockets.
   - Both modules use WebSockets for communication, but the distinction lies in whether the application is a WebAssembly application or a standalone application.
   - Most of the code is shared between the two modules, with the primary difference being the way the connection is established.

2. Challenges and Solutions:
   - The team encountered challenges in ensuring that the client API works correctly with WebAssembly applications. They have made necessary changes to address these issues, resulting in a functional API that works well with WebAssembly.
   - The team discussed the need for a way to manage contract and delegate hashes within applications. They have implemented a solution that involves embedding hashes into the build script, which allows applications to reference contracts and delegates by their hashes. However, they are open to exploring more efficient solutions in the future.

3. Identity and Donations:
   - The team is exploring the idea of allowing users to reserve identities authenticated through donations. These identities would be created and endorsed with a signature indicating the donation amount and date.
   - The process would involve generating a public-private key pair in the user's browser, blinding the public key, and sending it to the server for signing. The signature is then sent back to the browser and unblinded.
   - The goal is to incentivize donations and provide users with a sense of value for their early contributions. The team believes that this approach could help with funding and address the issue of identity authentication within the network.

4. Commutative Monoids and State Consistency:
   - The team discussed the concept of a "commutative monoid" as a key component of the Locutus project. Commutative monoids are mathematical structures that are useful for defining merge operations to achieve state consistency in distributed systems.
   - In the context of Locutus, commutative monoids are used to define how data is merged and reconciled across different nodes in the network. This is particularly important for ensuring data consistency in a decentralized environment where multiple nodes may independently update the same data.
   - What makes the team's approach novel is that the contract creator is responsible for defining the merge operation for its own data. This flexibility allows for the creation of a wide range of applications, each with its own user-defined merge operators tailored to the specific use case and data model.
   - By leveraging the properties of commutative monoids, the team aims to provide a robust and flexible framework for building distributed applications on Locutus. The goal is to ensure that data remains consistent across the network, even in the face of concurrent updates and network partitions.

5. Future Plans:
   - The team is focused on completing the development of the email app, which will serve as an early Minimum Viable Product (MVP) for Locutus.
   - They are also working on improving the project's website and exploring ways to attract more donors.
   - The team plans to shift their focus to the development of the node and network components of Locutus once the email app is completed.

Overall, the Freenet development team is making steady progress on the Locutus project. They are actively addressing challenges, exploring innovative solutions, and working towards the completion of key components.

**Comment by ple1n** [comment]
Would it be *meaningful* to build a decentralized embedding database / vector database ? P2P protocols of present and past are merely data transports, Ex. text chat, file sharing, social platform (text chat in a different way). It may be meaningless if possible developments in AI make vector databases useless, or fundamentally AI (part of it, like, the database) can't be efficiently/realistically decentralized.

I'm wondering whether there is anything that can be done on the protocol side, to make use of the opensource models, and apply the bit-torrent culture to the modern computing that runs on AI.

**Reply by sanity** [reply]
You could build an embedding/vector database on Locutus, you'd need to index the database using a [locality sensitive hash](https://en.wikipedia.org/wiki/Locality-sensitive_hashing) of the vectors spread across contracts, a different contract for each hash. 

The contract's state would contain all vectors with that hash, which the application would download to find the closest.
